### PR TITLE
o/snapstate: improve snapshot iteration

### PR DESCRIPTION
**Pull request by stolowski**
Improve snapshot Iter method to ignore files that do not end with .zip (it currently treats all files as .zips and logs errors for any random file in snapshots directory) or whose set id deducted from filename doesn't match set id stored inside metadata.
The latter change prepares snapd for future changes in that area where only the set id in the filename will matter.

### CODE REVIEW

File: overlord/snapshotstate/backend/backend_test.go
Comment (mvo5) [inline overlord/snapshotstate/backend/backend_test.go:377]: This should be "readNames > 1", the code in backend.go is a bit confusing, see my comment below about using a counter for "calledF". dirNames() will be called in a loop but all the relevant names are already returned in the first loop so no need to run it twice.

File: overlord/snapshotstate/backend/backend_test.go
Comment (mvo5) [inline overlord/snapshotstate/backend/backend_test.go:415]: This should be "readNames > 1", the code inbackend.go is a bit confusing, see my comment above/below about using a counter for "calledF". dirNames() will be called in a loop but all the relevant names are already returned in the first loop so no need to run it twice.

**Review by mvo5 — COMMENTED** [review]
Thanks, looks good! Some small suggestions in the testing section :)

File: overlord/snapshotstate/backend/backend.go
Comment (bboozzoo) [inline overlord/snapshotstate/backend/backend.go:190]: Nitpick, but this will work for `1_.zip` too. Maybe we should expect the parts[1] to conform to snap name? We could drop the extension and then exploy naming.ValidateSnap().

File: overlord/snapshotstate/backend/backend_test.go
Comment (bboozzoo) [reply overlord/snapshotstate/backend/backend_test.go:392]: A counter won't hurt probably.

File: overlord/snapshotstate/backend/backend_test.go
Comment (bboozzoo) [reply overlord/snapshotstate/backend/backend_test.go:377]: Yes, `readNames > 1`, the code loops until it gets an error from dirNames(), so to mimic os.Readdirnames, we should return the list on the first call, and then io.EOF on the second.

File: overlord/snapshotstate/backend/backend_test.go
Comment (bboozzoo) [inline overlord/snapshotstate/backend/backend_test.go:423]: Can you add `43_.zip` to the list?

File: overlord/snapshotstate/backend/backend.go
Comment (stolowski) [reply overlord/snapshotstate/backend/backend.go:190]: Right. I'm not sure we should do that, but I'm open to discuss. FWTW we are mostly interested in filtering out non-zip files and files where we cannot get set-id from the beginning of the filename. The snapshots directory is fully managed and owned by snapd so we don't expect random files there, the general idea is to make it future-proof in case we want to add something new there. At the moment it is not even possible to add a subdirectory there without triggering errors during iteration ;)

File: overlord/snapshotstate/backend/backend_test.go
Comment (stolowski) [reply overlord/snapshotstate/backend/backend_test.go:420]: Per earlier comments this would be considered valid; instead I've added this as a testcase to TestIsSnapshotFilename (with comment).

**Comment by stolowski** [comment]
@mvo5 @bboozzoo I've addressed your comments re tests tweaks/fixes, thanks for catching.

File: overlord/snapshotstate/backend/backend.go
Comment (pedronis) [inline overlord/snapshotstate/backend/backend.go:122]: the code used the spelling "set id", no hyphen, let's be consistent

Comment (pedronis) [reply overlord/snapshotstate/backend/backend.go:122]: this should be a Noticef I think

File: overlord/snapshotstate/backend/backend.go
Comment (pedronis) [reply overlord/snapshotstate/backend/backend.go:190]: the point of this is mostly filtering out non zips, and finding the set id from the name.  I wouldn't got the full ValidateSnap route, OTOH it might make sense to ignore things where len(parts[1]) == 0, we expect something to be there

File: overlord/snapshotstate/backend/backend_test.go
Comment (pedronis) [inline overlord/snapshotstate/backend/backend_test.go:149]: we might want to make this one false as well, as mentioned above

**Review by pedronis — APPROVED** [review]
thank you

File: overlord/snapshotstate/backend/backend.go
Comment (mvo5) [reply overlord/snapshotstate/backend/backend.go:122]: Heh, I think we asked him to change it. Oh well, it can be changed back, no big deal. But I really think we should not make this a warning, I would remove the "XXX"

**Review by mvo5 — APPROVED** [review]
LGTM once Samueles points are addressed.

**Comment by mvo5** [comment]
I push the trivial changes, once green(ish) this can be merged

**Pull request merged**
